File: src/test/java/com/binance/api/domain/event/CandlestickEventDeserializerTest.java
package com.binance.api.domain.event;

import com.binance.api.client.domain.event.CandlestickEvent;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.Test;

import java.io.IOException;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

/**
 * Tests that JSON responses from ta candlestick event are converted to the appropriate <code>CandlestickEvent</code> object.
 */
public class CandlestickEventDeserializerTest {

  @Test
  public void testCandlestickEventDeserializer() {
    String candlestickEventJson = "{\n" +
        "        \"e\": \"kline\",\n" +
        "        \"E\": 1,\n" +
        "        \"s\": \"ETHBTC\",\n" +
        "        \"k\": {\n" +
        "        \"t\": 1499404860000,\n" +
        "        \"T\": 1499404919999,\n" +
        "        \"s\": \"ETHBTC\", \n" +
        "        \"i\": \"1m\",\n" +
        "        \"f\": 77462, \n" +
        "        \"L\": 77465, \n" +
        "        \"o\": \"0.10278577\", \n" +
        "        \"c\": \"0.10278645\", \n" +
        "        \"h\": \"0.10278712\", \n" +
        "        \"l\": \"0.10278518\", \n" +
        "        \"v\": \"17.47929838\", \n" +
        "        \"n\": 4, \n" +
        "        \"x\": false, \n" +
        "        \"q\": \"1.79662878\", \n" +
        "        \"V\": \"2.34879839\", \n" +
        "        \"Q\": \"0.24142166\", \n" +
        "        \"B\": \"13279784.01349473\"\n" +
        "      }}";
    ObjectMapper mapper = new ObjectMapper();
    try {
      CandlestickEvent candlestickEvent = mapper.readValue(candlestickEventJson, CandlestickEvent.class);
      assertEquals(candlestickEvent.eventType, "kline");
      assertEquals(candlestickEvent.eventTime, 1L);
      assertEquals(candlestickEvent.symbol, "ETHBTC");

      assertEquals((long)candlestickEvent.openTime, 1499404860000L);
      assertEquals(candlestickEvent.open, "0.10278577");
      assertEquals(candlestickEvent.high, "0.10278712");
      assertEquals(candlestickEvent.low, "0.10278518");
      assertEquals(candlestickEvent.close, "0.10278645");
      assertEquals(candlestickEvent.volume, "17.47929838");
      assertEquals((long)candlestickEvent.closeTime, 1499404919999L);
      assertEquals(candlestickEvent.intervalId, "1m");
      assertEquals((long)candlestickEvent.firstTradeId, 77462L);
      assertEquals((long)candlestickEvent.lastTradeId, 77465L);
      assertEquals(candlestickEvent.quoteAssetVolume, "1.79662878");
      assertEquals((long)candlestickEvent.numberOfTrades, 4L);
      assertEquals(candlestickEvent.takerBuyBaseAssetVolume, "2.34879839");
      assertEquals(candlestickEvent.takerBuyQuoteAssetVolume, "0.24142166");
      assertEquals(candlestickEvent.isBarFinal, false);
    } catch (IOException e) {
      fail(e.getMessage());
    }
  }
}

File: src/test/java/com/binance/api/domain/event/CandlestickEventDeserializerTest.scala
package com.binance.api.domain.event

import com.binance.api.client.domain.event.{CandlestickDetailed, CandlestickEvent}
import com.binance.api.client.json.Decoders._
import io.circe._
import org.junit.Assert.assertEquals
import org.junit.Test

/**
  * Tests that JSON responses from ta candlestick event are converted to the appropriate <code>CandlestickEvent</code> object.
  */
class CandlestickEventDeserializerTest {
  @Test def testCandlestickEventDeserializer(): Unit = {
    val json = s"""{
        "e": "kline",
        "E": 1,
        "s": "ETHBTC",
        "k": {
        "t": 1499404860000,
        "T": 1499404919999,
        "s": "ETHBTC", 
        "i": "1m",
        "f": 77462, 
        "L": 77465, 
        "o": "0.10278577", 
        "c": "0.10278645", 
        "h": "0.10278712", 
        "l": "0.10278518", 
        "v": "17.47929838", 
        "n": 4, 
        "x": false, 
        "q": "1.79662878", 
        "V": "2.34879839", 
        "Q": "0.24142166", 
        "B": "13279784.01349473"
      }}"""
    assertEquals(
      Right(
        CandlestickEvent(
          eventType = "kline",
          eventTime = 1,
          symbol = "ETHBTC",
          candlestick = CandlestickDetailed(
            openTime = 1499404860000L,
            open = "0.10278577",
            high = "0.10278712",
            low = "0.10278518",
            close = "0.10278645",
            volume = "17.47929838",
            closeTime = 1499404919999L,
            intervalId = "1m",
            firstTradeId = 77462,
            lastTradeId = 77465,
            quoteAssetVolume = "1.79662878",
            numberOfTrades = 4,
            takerBuyBaseAssetVolume = "2.34879839",
            takerBuyQuoteAssetVolume = "0.24142166",
            isBarFinal = false
          )
        )
      ),
      parser.decode[CandlestickEvent](json)
    )
  }
}
